lucetc/src/function.rs

```diff
@@ -302,4 +306,136 @@ impl<'a> FuncEnvironment for FuncInfo<'a> {
+        // Loop, BrIf, CallIndirect), update the wasm instruction counter and:
+        // * if the operation introduces a new scope (If, Block, Loop), push a new 0 on the
+        // stack corresponding with that frame.
+        // * if the operation does not introduce a new scope (Else, Call, CallIndirect, BrIf),
+        // reset the top of stack to 0
+        // * if the operation completes a scope (End), pop the top of the stack and reset the new
+        // top of stack to 0
+        // * this leaves no special behavior for Unreachable and Return. This is acceptable as they
+        // are always followed by an End and are either about to trap, or return from a function.
+        // * Unreachable is either the end of VM execution and we are off by one instruction, or,
+        // is about to dispatch to an exception handler, which we should account for out of band
+        // anyway (exception dispatch is much more expensive than a single wasm op)
+        // * Return corresponds to exactly one function call, so we can count it by resetting the
+        // stack to 1 at return of a function.
+
+        fn flush_counter(environ: &mut FuncInfo, builder: &mut FunctionBuilder) {
+            if environ.op_offsets.last() == Some(&0) {
+                return;
+            }
+            let instr_count_offset: ir::immediates::Offset32 =
+                (-(std::mem::size_of::<InstanceRuntimeData>() as i32)
+                    + offset_of!(InstanceRuntimeData, instruction_count) as i32)
+                    .into();
+            let vmctx_gv = environ.get_vmctx(builder.func);
+            let addr = builder.ins().global_value(environ.pointer_type(), vmctx_gv);
+            let flags = ir::MemFlags::trusted();
+            let cur_instr_count =
+                builder
+                    .ins()
+                    .load(ir::types::I64, flags, addr, instr_count_offset);
+            let update_const = builder.ins().iconst(
+                ir::types::I64,
+                i64::from(*environ.op_offsets.last().unwrap()),
+            );
+            let new_instr_count = builder.ins().iadd(cur_instr_count, update_const.into());
+            builder
+                .ins()
+                .store(flags, new_instr_count, addr, instr_count_offset);
+            *environ.op_offsets.last_mut().unwrap() = 0;
+        };
+
+        // Update the instruction counter, if necessary
+        let op_cost = match op {
+            // Opening a scope is a syntactic operation, and free.
+            Operator::Block { .. } |
+            // These do not add counts, see above comment about return/unreachable
+            Operator::Unreachable |
+            Operator::Return => 0,
+            // Call is quick
+            Operator::Call { .. } => 1,
+            // but indirect calls take some extra work to validate at runtime
+            Operator::CallIndirect { .. } => 2,
```

Comment:
Most wasm operations are either syntactic (like Block and End) or have some close correspondence with a single-ish machine instruction. `CallIndirect` and `BrTable` involve several operations when translated to clif and later native code, so I count them twice to represent that. Totally arbitrary otherwise.